### PR TITLE
#bugfix defined at https://github.com/jcabi/jcabi-w3c/issues/16

### PR DESCRIPTION
**Pull request by jrdalpra**
@dmarkov  PR that refers to issue at https://github.com/jcabi/jcabi-w3c/issues/16

- added new unit test that asserts that an invalid html must 
  return valid
  as false, contains at least one error and and one warning
- added new integration test that validates direct at W3C an invalid
  html and asserts that errors list is not empty

### CODE REVIEW

**Comment by dmarkov** [comment]
@jrdalpra Many thanks for the PR, let me find a reviewer for it

**Comment by dmarkov** [comment]
@dmzaytsev this pull request is for you, please review

**Comment by dmzaytsev** [comment]
@jrdalpra CI builds failed
let's fix it

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java [hunk 1]
```diff
@@ -38,7 +38,7 @@
 /**
  * Integration case for {@link DefaultHtmlValidator}.
  * @author Yegor Bugayenko (yegor@tpc2.com)
- * @version $Id$
+ * @version $Id: c8f79647cb1da6ce4802f3a0f7bdef46301061c1 $
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: @jrdalpra no need to modify this line
please revert the change

**Comment by dmzaytsev** [comment]
@jrdalpra according the quality rules see http://at.teamed.io/qa.html
the pull request description should contains a link to the original ticket it is related to.

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 2]
```diff
@@ -29,98 +29,118 @@
...
+import com.jcabi.http.mock.MkAnswer;
+import com.jcabi.http.mock.MkContainer;
+import com.jcabi.http.mock.MkGrizzlyContainer;
+
 /**
  * Test case for {@link DefaultHtmlValidator}.
+ * 
  * @author Yegor Bugayenko (yegor@tpc2.com)
- * @version $Id$
+ * @version $Id: 802502850df222023e7cfbb6932b8e9ef0684a2c $
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra no need to modify this line

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java [hunk 3]
```diff
@@ -62,5 +62,16 @@ public void validatesHtmlDocument() throws Exception {
             Matchers.empty()
         );
     }
-
+    
+    @Test
+    @RetryOnFailure(verbose = false)
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: @jrdalpra why you need this annotation here?

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java [hunk 4]
```diff
@@ -62,5 +62,16 @@ public void validatesHtmlDocument() throws Exception {
...
+    public void validatesInvalidHtmlDocument() throws Exception {
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: @jrdalpra javadoc for the test method is missed

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 5]
```diff
@@ -29,98 +29,118 @@
...
+	 * DefaultHtmlValidator can validate HTML document.
+	 * 
+	 * @throws Exception
+	 *             If something goes wrong inside
+	 */
+	@Test
+	public void validatesHtmlDocument() throws Exception {
+		final MkContainer container = new MkGrizzlyContainer()
+				.next(new MkAnswer.Simple(
+						StringUtils
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra the indentation step should be  4 spaces here and below

**Comment by dmzaytsev** [comment]
@jrdalpra a few comments above
and a lot checkstyle complains in CI build

**Comment by dmzaytsev** [comment]
@jrdalpra why did you close it?

**Comment by jrdalpra** [comment]
@dmzaytsev to make a new one.

I need to fix the problems here? Sorry, but I thought was better close this and open a new one.

**Comment by jrdalpra** [comment]
@dmarkov 
@dmzaytsev here is the new PR https://github.com/jcabi/jcabi-w3c/pull/24

**Comment by dmzaytsev** [comment]
@jrdalpra You just needed to push commits into same branch

**Comment by jrdalpra** [comment]
@dmzaytsev  Sorry. 

So what do I do now? Reopen this one or use the new PR?

**Comment by dmzaytsev** [comment]
@jrdalpra Let's reopen  this one

**Comment by jrdalpra** [comment]
@dmzaytsev I think I don't have permission to reopen the PR. Can you reopen it?
image

**Comment by dmzaytsev** [comment]
@jrdalpra that's why you already have an open pull request from same branch
try to close a new PR and then try to reopen this one

**Comment by jrdalpra** [comment]
@dmzaytsev done. Sorry!

**Comment by dmzaytsev** [comment]
@jrdalpra 
> the pull request description should contains a link to the original ticket it is related to.

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java [hunk 6]
```diff
@@ -45,21 +46,44 @@
...
-            Matchers.empty()
+            Matchers.not(Matchers.empty())
+        );
+    }
+
+    /**
+     * Creates a valid html.
+     * @return Valid html content
+     */
+    private String validHtml() {
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: @jrdalpra this method used only once
Why you introduced this one?

Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: @jrdalpra lets keep `validatesHtmlDocument()` unchanged

Comment (jrdalpra) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: I think it's better to read the code. Don't you think?

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 7]
```diff
@@ -91,13 +114,13 @@ public void validatesHtmlDocument() throws Exception {
...
         final Set<Integer> responses = new HashSet<Integer>(
-            Arrays.asList(
-                HttpURLConnection.HTTP_INTERNAL_ERROR,
-                HttpURLConnection.HTTP_NOT_IMPLEMENTED,
-                HttpURLConnection.HTTP_BAD_GATEWAY,
-                HttpURLConnection.HTTP_UNAVAILABLE,
-                HttpURLConnection.HTTP_GATEWAY_TIMEOUT,
-                HttpURLConnection.HTTP_VERSION
+                        Arrays.asList(
+                                      HttpURLConnection.HTTP_INTERNAL_ERROR,
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra lets keep the indentation here and below unchanged
we use 4 spaces usually

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java (hunk 6 above)
Comment (jrdalpra) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: Maybe the checkstyle rule to use 'this' keyword makes it no so redable.

Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: @jrdalpra we should avoid once-used methods and variables

Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: @jrdalpra also please address your messages
it's the quality requirements

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 8]
```diff
@@ -105,22 +128,58 @@ public void throwsIOExceptionWhenValidationServerErrorOccurred()
...
     }
 
+    /**
+     * Use a file to build the request.
+     * @return Request inside the file.
+     * @throws IOException if something goes wrong.
+     */
+    private String invalidHtmlResponse() throws IOException {
+        final InputStream file = DefaultHtmlValidator.class
+                        .getResourceAsStream("invalid-html-response.xml");
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra why we need so big invalid html?
I think we use html as the ticket reporter suggested

**Comment by dmzaytsev** [comment]
@jrdalpra please see a few comments above

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java (hunk 6 above)
Comment (jrdalpra) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: @dmzaytsev do you prefer a non readable code than create a simple method or variable?

Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: @jrdalpra first I prefer don't touch the code that does not refer to the ticket :)
You should create a new test to reproduce the issue and than fix it, if you have a time budget or create a puzzle otherwise.

yes if the code really unreadble, we can a helper methods.
but in this case the code was readable enough I think

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java (hunk 8 above)
Comment (jrdalpra) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @dmzaytsev it's not the request but the W3C response (mock). This file contains the real W3C response for "this is and invalid html" validation request.

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java (hunk 6 above)
Comment (jrdalpra) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: @dmzaytsev I always prefer to make code more readable. Machines are cheaper than developers. :) 
But I'll what you're asking for.

**Comment by dmzaytsev** [comment]
@jrdalpra all messages in the ticket (PR in this case) should start with a name of a user 
please fix the messages

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java (hunk 8 above)
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra got it

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java (hunk 7 above)
Comment (jrdalpra) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @dmzaytsev  done

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java (hunk 8 above)
Comment (jrdalpra) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @dmzaytsev should I keep that way or change it?

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java (hunk 6 above)
Comment (jrdalpra) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: @dmzaytsev done

**Comment by jrdalpra** [comment]
@dmzaytsev all done?

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java (hunk 8 above)
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra yes, keep that way

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 9]
```diff
@@ -105,22 +128,58 @@ public void throwsIOExceptionWhenValidationServerErrorOccurred()
             MkContainer container = null;
             try {
                 container = new MkGrizzlyContainer().next(
-                    new MkAnswer.Simple(status)
+                                new MkAnswer.Simple(status)
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra no needs to change indentation here
let's use 4 spaces instead 8

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 10]
```diff
@@ -105,22 +128,58 @@ public void throwsIOExceptionWhenValidationServerErrorOccurred()
...
                 ).start();
                 new DefaultHtmlValidator(container.home())
-                    .validate("<html></html>");
+                                .validate("<html></html>");
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra same here

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java [hunk 11]
```diff
@@ -63,4 +65,18 @@ public void validatesHtmlDocument() throws Exception {
...
 
+    /**
+     * DefaultHtmlValidator can validate invalid HTML document.
+     *
+     * @throws Exception If something goes wrong inside
+     */
+    @Test
+    public void validatesInvalidHtmlDocument() throws Exception {
+        MatcherAssert.assertThat(
+                        ValidatorBuilder.HTML.validate(
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorITCase.java]: @jrdalpra 4 spaces

**Comment by dmzaytsev** [comment]
@jrdalpra just invalid indentation above

**Comment by dmzaytsev** [comment]
@jrdalpra where did you fix the issue?
or you can't reproduce it?

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java (hunk 10 above)
Comment (jrdalpra) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @dmzaytsev done.

Eclipse formatter is killing my profits. :)

**Comment by jrdalpra** [comment]
@dmzaytsev sorry, can you be more specific?

**Comment by dmzaytsev** [comment]
@jrdalpra nevermind, my mistake

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 12]
```diff
@@ -78,12 +72,41 @@ public void validatesHtmlDocument() throws Exception {
...
+    @Test
+    public void validateInvalidHtml() throws Exception {
+        final MkContainer container = new MkGrizzlyContainer().next(
+                        new MkAnswer.Simple(this.invalidHtmlResponse())
+        ).start();
+        final Validator validator = new DefaultHtmlValidator(container.home());
+        final ValidationResponse response = validator
+            .validate("this is an invalid html");
+        container.stop();
+        MatcherAssert.assertThat(
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra 
```
MatcherAssert.assertThat(
                        "Validity must be invalid!",
                        !response.valid()
        );
```
should be 
```
MatcherAssert.assertThat(
    "Validity must be invalid!",
    !response.valid()
);
```

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 13]
```diff
@@ -78,12 +72,41 @@ public void validatesHtmlDocument() throws Exception {
...
+                        "Validity must be invalid!",
+                        !response.valid()
+        );
+        MatcherAssert.assertThat(
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra 
```
MatcherAssert.assertThat(
                        "Must has at least one error",
                        response.errors(), this.withoutDefects()
        );
```
should be 
```
MatcherAssert.assertThat(
    "Must has at least one error",
    response.errors(), 
    this.withoutDefects()
);
```

same below

I wonder why checkstyle doesn't report about it

Comment (jrdalpra) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @dmzaytsev done.

Me too.

I'm the only one that uses Eclipse at teamed.io?

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 14]
```diff
@@ -115,12 +139,48 @@ public void throwsIOExceptionWhenValidationServerErrorOccurred()
...
+                .join(
+                      "<env:Envelope",
+                      " xmlns:env='http://www.w3.org/2003/05/soap-envelope'>",
+                      "<env:Body><m:markupvalidationresponse",
+                      " xmlns:m='http://www.w3.org/2005/10/markup-validator'>",
+                      "<m:validity>true</m:validity>",
+                      "<m:checkedby>W3C</m:checkedby>",
+                      "<m:doctype>text/html</m:doctype>",
+                      "<m:charset>UTF-8</m:charset>",
+                      "</m:markupvalidationresponse></env:Body></env:Envelope>"
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra 
```
    return StringUtils.join(
        "<env:Envelope",
        " xmlns:env='http://www.w3.org/2003/05/soap-envelope'>",
        "<env:Body><m:markupvalidationresponse",
        " xmlns:m='http://www.w3.org/2005/10/markup-validator'>",
        "<m:validity>true</m:validity>",
        "<m:checkedby>W3C</m:checkedby>",
        "<m:doctype>text/html</m:doctype>",
        "<m:charset>UTF-8</m:charset>",
        "</m:markupvalidationresponse></env:Body></env:Envelope>"
);
```

Comment (jrdalpra) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @dmzaytsev done

**Comment by dmzaytsev** [comment]
@jrdalpra one more
https://github.com/jcabi/jcabi-w3c/pull/21/files#diff-f66c0770fbfd0c1e74cf49800f02d73aR76

and we are ready to merge

**Comment by jrdalpra** [comment]
@dmzaytsev done

**Comment by dmzaytsev** [comment]
@jrdalpra thank you!

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by jrdalpra** [comment]
@dmzaytsev thanks for your patience

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 15]
```diff
@@ -78,12 +72,42 @@ public void validatesHtmlDocument() throws Exception {
     }
 
     /**
+     * Test if {@link DefaultHtmlValidator} validades invalid html.
+     * @throws Exception If something goes wrong inside
+     */
+    @Test
+    public void validateInvalidHtml() throws Exception {
+        final MkContainer container = new MkGrizzlyContainer().next(
+                        new MkAnswer.Simple(this.invalidHtmlResponse())
```
Comment (yegor256) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra why so many indentation spaces?? we're using 4 spaces, always

**Comment by yegor256** [comment]
@jrdalpra looks good, but why are you using so many indentation spaces sometimes? should be 4, always.

**Comment by jrdalpra** [comment]
@yegor256 eclipse formatted the code. I fixed some parts but I think it
passed.

Em sexta-feira, 11 de dezembro de 2015, Yegor Bugayenko <
notifications@github.com> escreveu:

> @jrdalpra <https://github.com/jrdalpra> looks good, but why are you using
> so many indentation spaces sometimes? should be 4, always.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jcabi/jcabi-w3c/pull/21#issuecomment-164083320>.
>


-- 
José Volmei Dal Prá Junior - (49) 9112 9359
Bacharel em Ciência da Computação - UnoChapecó
Pós-Graduado em Desenvolvimento de Software Com Java - Unoesc Xxe
Consultor Java e TotalCross
https://github.com/jrdalpra
http://www.kombo.com.br/meucurriculo/jrdalpra
http://www.linkedin.com/in/jrdalpra
msn: jr_dalpra@hotmail.com
skype: jrdalpra
e-mail: jrdalpra@gmail.com

**Comment by jrdalpra** [comment]
@dmzaytsev @yegor256 what's the PR status?

**Comment by dmzaytsev** [comment]
@jrdalpra I see a comment from @yegor256 above
you didn't fix it yet

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java (hunk 15 above)
Comment (jrdalpra) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @yegor256 @dmzaytsev done

**Comment by jrdalpra** [comment]
@dmzaytsev @yegor256 any news?

**Comment by jrdalpra** [comment]
@dmzaytsev @yegor256 Can you give me a feedback? I want to get more tasks and I have de limit of 2. Thanks.

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 16]
```diff
@@ -107,18 +138,72 @@ public void throwsIOExceptionWhenValidationServerErrorOccurred()
...
+    }
+
+    /**
+     * Use a file to build the request.
+     * @return Request inside the file.
+     * @throws IOException if something goes wrong.
+     */
+    private String invalidHtmlResponse() throws IOException {
+        final InputStream file = DefaultHtmlValidator.class.getResourceAsStream(
+               "invalid-html-response.xml"
```
Comment (yegor256) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra we always use 4 spaces for indentation... why 6 here?

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 17]
```diff
@@ -107,18 +138,72 @@ public void throwsIOExceptionWhenValidationServerErrorOccurred()
...
+        );
+        final String xml = IOUtils.toString(
```
Comment (yegor256) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra can't you put all this in one line? why making three out of this?

File: src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java [hunk 18]
```diff
@@ -107,18 +138,72 @@ public void throwsIOExceptionWhenValidationServerErrorOccurred()
...
+            file
+        );
+        IOUtils.closeQuietly(
```
Comment (yegor256) [inline src/test/java/com/jcabi/w3c/DefaultHtmlValidatorTest.java]: @jrdalpra same here, why three lines instead of one?

**Comment by yegor256** [comment]
@jrdalpra a few small comments from me above

**Comment by jrdalpra** [comment]
@yegor256 @dmzaytsev done

Qulice is not getting this formatting problems

**Comment by jrdalpra** [comment]
done

2015-12-17 19:20 GMT-02:00 Yegor Bugayenko <notifications@github.com>:

> @jrdalpra <https://github.com/jrdalpra> a few small comments from me above
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jcabi/jcabi-w3c/pull/21#issuecomment-165584654>.
>



-- 
José Volmei Dal Prá Junior - (49) 9112 9359
Bacharel em Ciência da Computação - UnoChapecó
Pós-Graduado em Desenvolvimento de Software Com Java - Unoesc Xxe
Consultor Java e TotalCross
https://github.com/jrdalpra
http://www.kombo.com.br/meucurriculo/jrdalpra
http://www.linkedin.com/in/jrdalpra
msn: jr_dalpra@hotmail.com
skype: jrdalpra
e-mail: jrdalpra@gmail.com

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by dmarkov** [comment]
@dmzaytsev **52 mins** was added to your account, many thanks for your contribution (payment 72855351)! The task took 240 hours and 56 mins.. review comments (c=37) added as a bonus. +52 to your rating, your total score is [+1090](http://www.netbout.com/b/35166?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy

**Comment by jrdalpra** [comment]
@dmarkov @yegor256 @dmzaytsev PR merged and task still open at netbout. How can we solve it?

**Comment by jrdalpra** [comment]
@dmarkov @yegor256 @dmzaytsev PR merged and task still open at netbout. How can we solve it?

**Comment by jrdalpra** [comment]
closes #16
